src/gluonts/itertools.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from typing import Iterable, Iterator, List, TypeVar
+import itertools
+import random
+
+T = TypeVar("T")
+
+
+def cycle(it):
+    """Like `itertools.cycle`, but does not store the data."""
+
+    while True:
+        yield from it
+
+
+def take(iterable: Iterable[T], n: int) -> Iterator[T]:
+    """Returns up to `n` elements from `iterable`.
+
+    This is similar to xs[:n], except that it works on `Iterable`s and possibly
+    consumes the given `iterable`.
+
+    >>> list(take(range(10), 5))
+    [0, 1, 2, 3, 4]
+    """
+    return itertools.islice(iterable, n)
+
+
+def batcher(iterable: Iterable[T], batch_size: int) -> Iterator[List[T]]:
+    """Groups elements from `iterable` into batches of size `batch_size`.
+
+    >>> list(batcher("ABCDEFG", 3))
+    [['A', 'B', 'C'], ['D', 'E', 'F'], ['G']]
+
+    Unlike the grouper proposed in the documentation of itertools, `batcher`
+    doesn't fill up missing values.
+    """
+    it: Iterator[T] = iter(iterable)
+
+    def get_batch():
+        return list(take(it, batch_size))
+
+    # has an empty list so that we have a 2D array for sure
+    return iter(get_batch, [])
+
+
+class _CachedIterable(Iterable):
```

Comment:
Couldn't we just do `class cache`?

Reply:
we could, I just didn't want to break the sacred tradition of upper-case initials for class names

Reply:
The exception that proves the rule!